documentation/development/architecture.md

```diff
@@ -0,0 +1,56 @@
+# Architecture of the Git Town code base
+
+## Source code
+
+Git Town is written in [Go](https://golang.org).
+The source code is located in the [src](../../src) folder.
+
+### CLI wrapper
+
+Git Town uses the [Cobra](https://github.com/spf13/cobra) CLI framework.
+[src/cmd](../../src/cmd) contains the commands that Git Town understands.
+
+### Steps
+
+The individual steps that each Git Town command executes are expressed via the
+[command pattern](https://en.wikipedia.org/wiki/Command_pattern).
+This allows allow for robust and fully automated `--continue`, `--abort`,
+and `--undo` functionality.
+To distinguish the command-pattern commands from the Git Town commands in [src/cmd](../../src/cmd),
+we'll call the former `steps` from now on.
+Steps are in [src/steps](../../src/steps)
+and implement the individual steps that each Git Town command performs,
+like for example [changing to a different Git branch](../../src/steps/checkout_branch_step.go)
+or [pulling down updates for the current branch](../../src/steps/pull_branch_step.go).
+
+Each of the possible steps is a Go struct that has a `Run` method to execute the step
+as well as a `CreateUndoStepBeforeRun` method
+that returns a step that performs the inverse operation.
```

Comment:
I think a lot of this section already lives here: https://github.com/Originate/git-town/blob/master/documentation/development/steps_list.md

Reply:
Good point, I have merged my description here with steps_list.md